javascript/ql/src/semmle/javascript/explore/CallGraph.qll

```diff
@@ -0,0 +1,85 @@
+/**
+ * Provides predicates for visualizing the call paths leading to or from a specific function.
+ *
+ * It defines three predicates: `callEdge`, `isStartOfCallPath` and `isEndOfCallPath`,
+ * as well as the `nodes` and `edges` predicates needed for a path problem query.
+ *
+ * To use this library, make sure the query has `@kind path-problem`
+ * and selects columns appropriate for a path problem query.
+ * For example:
+ * ```
+ * import javascript
+ * import semmle.javascript.explore.CallGraph
+ * import DataFlow
+ *
+ * from InvokeNode invoke, FunctionNode function
+ * where callEdge*(invoke, function)
+ *   and isStartOfCallPath(invoke)
+ *   and function.getName() = "targetFunction"
+ * select invoke, invoke, function, "Call path to 'targetFunction'"
+ * ```
+ */
+import javascript
+private import DataFlow
+
+/**
+ * Holds if `pred -> succ` is an edge in the call graph.
+ *
+ * There are edges from calls to their callees,
+ * and from functions to their contained calls and in some cases
+ * their inner functions.
```

Comment:
That last bit is interesting; why do we want those edges?

Reply:
To handle callbacks passed to `forEach` where we don't see the exact call site.

So in a sense, a call edge `A -> B` can be read as `A causes B to be called`.

Even when it's not technically called in the same stack frame (e.g. through `addEventListener`) I still found this to be useful.

Reply:
OK, that makes sense. Could you expand the documentation slightly to explain this?

Also, I wonder whether it would make sense to distinguish the two kinds of edges by giving them different labels?